### DISCUSSION TITLE
Drawing an image loaded via include_bytes!()?

### DISCUSSION THREAD

**Discussion by 4JX** [fltk-rs/fltk-rs#998] [Q&A] [answered]
I have the following snippet:
```rs
let icon_bytes = include_bytes!("../../res/trayIcon.png");
let mut icon = fltk::image::PngImage::from_data(icon_bytes).unwrap();
icon.draw(0, 0, 100, 100);
```
Which results in:

```sh
XRequest.139: BadDrawable (invalid Pixmap or Window parameter) 0x5a00083
XRequest.139: RenderBadPicture (invalid Picture parameter) 0x5a00089
XRequest.139: RenderBadPicture (invalid Picture parameter) 0x5a00089
XRequest.139: RenderBadPicture (invalid Picture parameter) 0x5a00089
```

The image is a valid file and the path is correct. It is properly loaded into icon_bytes, so what am I missing? I also tried an SVG with `include_str!()` to no avail.

**Comment by MoAlyousef** [comment, marked as answer]
icon.draw() would only work inside a widget's draw() method.
```rust
some_widget.draw(move |w| {
    icon.draw(args);
});
```

**Reply by 4JX** [reply]
So for example, a frame should suffice? I'm kind of confused on the use-case for each of the groups.

**Reply by MoAlyousef** [reply]
Yes it should. Widgets are grouped by the traits they implement, GroupExt widgets can parent other widgets and some also offer layouts etc. InputExt widgets basically take textual input ...etc. A frame only implements basic widget traits (WidgetExt and WidgetBase), it functions as a label, spacer, image holder etc.

**Reply by 4JX** [reply]
Hmm, the following does not work, but the frame does show up on green on the top-left corner. I seem to be missing something obvious 😅.

```rs
pub fn message(width: i32, height: i32, message: &str) {
	let width_center = width / 2;
	let height_center = height / 2;
	let window_x = screen_center().0 - width_center;
	let window_y = screen_center().1 - height_center;
	let margin = 20;
	let button_width = 100;
	let button_height = 40;

	let mut window = Window::new(window_x, window_y, width, height, "Message");
	window.set_color(Color::from_u32(Colors::DarkGray as u32));

	let mut buffer = fltk::text::TextBuffer::default();
	buffer.set_text(message);

	let mut display = fltk::text::TextDisplay::new(margin, margin, width - margin * 2, height - margin * 2 - button_height, "");
	display.set_buffer(buffer);
	display.set_color(Color::from_u32(Colors::DarkGray as u32));
	display.set_frame(FrameType::FlatBox);
	display.set_text_color(Color::from_u32(Colors::White as u32));

	let mut button = Button::new(width_center - button_height / 2, height - margin - button_height, button_width, button_height, "Close");
	button.set_color(Color::from_u32(Colors::Gray as u32));
	button.set_label_color(Color::White);
	button.set_frame(FrameType::BorderBox);

	let icon_bytes = include_bytes!("../../res/trayIcon.png");
	let mut icon = fltk::image::PngImage::from_data(icon_bytes).unwrap();
	let icon_str = include_str!("../../res/trayIcon.svg");
	let mut icon_svg = fltk::image::SvgImage::from_data(icon_str).unwrap();

	let mut image_frame = fltk::frame::Frame::new(0, 0, 100, 100, "");
	image_frame.set_frame(FrameType::FlatBox);
	image_frame.set_color(Color::Green);
	image_frame.draw({
		move |_frame| {
			icon.draw(0, 0, 100, 100);
			icon_svg.draw(0, 0, 100, 100);
		}
	});

	window.end();
	window.show();

	button.set_callback({
		move |_but| {
			window.hide();
		}
	});
}
```

**Reply by MoAlyousef** [reply]
The TextDisplay covers a large part of your image_frame

**Reply by 4JX** [reply]
> The TextDisplay covers a large part of your image_frame

Forgot to add, even without the TextDisplay being active.  https://ibb.co/xLS2VHz

**Reply by MoAlyousef** [reply]
If the png image is large and has alpha transparency, the actual part of the image might not be showing. You might have to scale the image down. You can try by passing a larger image_frame and drawing a larger part of the image.

**Reply by 4JX** [reply]
> If the png image is large and has alpha transparency, the actual part of the image might not be showing. You might have to scale the image down. You can try by passing a larger image_frame.

That worked, I was under the assumption that `image.draw()` would resize the image to the indicated size, my bad. On that note, is there a way to do such resizing on FLTK or would I have to do it externally and then pass the correctly sized image to FLTK?

*`image.scale()` seems to do the latter, thanks again!

**Reply by MoAlyousef** [reply]
You can use the scale method. Or if you set the frames image using set_image, you can use set_image_scaled().

**Reply by 4JX** [reply]
It would seem `set_image_scaled()` has an offset from the place where it should go, see: https://ibb.co/xH6FGcD

Red is a `TextDisplay` starting at the margin specified above, green is the frame that holds the image, and then there is the image itself which is slightly offset upwards. It's an SVG if that makes any difference, the file itself is not offset. (https://ibb.co/sPpFJbw)
